apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs

```diff
@@ -85,5 +92,120 @@ public void CloneUpdateWithParameters()
+            SpannerConnection connection = BuildSpannerConnection(spannerClientMock);
+            var queryOptions = new QueryOptions().WithOptimizerVersion(optimizerVersion);
+            connection.QueryOptions = queryOptions;
+
+            var command = connection.CreateSelectCommand("SELECT * FROM FOO");
+            using (var reader = command.ExecuteReader())
+            {
+                // Do nothing.
+            }
+
+            // Set the environment back.
+            Environment.SetEnvironmentVariable(optimizerVersionVariable, savedOptimizerVersion);
+        }
+
+        [Fact]
+        public void CommandHasOptimizerVersionSetOnCommand()
+        {
+            // Save existing value of environment variable.
+            const string optimizerVersionVariable = "SPANNER_OPTIMIZER_VERSION";
+            string savedOptimizerVersion = Environment.GetEnvironmentVariable(optimizerVersionVariable);
+            const string envOptimizerVersion = "2";
+            Environment.SetEnvironmentVariable(optimizerVersionVariable, envOptimizerVersion);
+
+            var cmdOptimizerVersion = "3";
+            // Optimizer version set at a command level has higher precedence
+            // than version set through the connection or the environment
+            // variable.
+            Mock<SpannerClient> spannerClientMock = SetupExecuteStreamingSql(cmdOptimizerVersion);
+
+            const string optimizerVersion = "1";
+            SpannerConnection connection = BuildSpannerConnection(spannerClientMock);
+            var queryOptions = new QueryOptions().WithOptimizerVersion(optimizerVersion);
+            connection.QueryOptions = queryOptions;
+
+            var command = connection.CreateSelectCommand("SELECT * FROM FOO");
+            command.QueryOptions = new QueryOptions().WithOptimizerVersion(cmdOptimizerVersion);
+            using (var reader = command.ExecuteReader())
+            {
+                // Do nothing.
+            }
+
+            // Set the environment back.
+            Environment.SetEnvironmentVariable(optimizerVersionVariable, savedOptimizerVersion);
+        }
+
+        private Mock<SpannerClient> SetupExecuteStreamingSql(string optimizerVersion)
+        {
+            Mock<SpannerClient> spannerClientMock = SpannerClientHelpers
+                .CreateMockClient(Logger.DefaultLogger, MockBehavior.Strict);
+            spannerClientMock
+                .SetupBatchCreateSessionsAsync()
+                .Setup(client => client.ExecuteStreamingSql(
+                    It.Is<ExecuteSqlRequest>(
+                        request => request.QueryOptions.OptimizerVersion == optimizerVersion),
+                    It.IsAny<CallSettings>()))
+                .Returns<ExecuteSqlRequest, CallSettings>((request, _) =>
+                {
+                    // Returning null since we can't return a SpannerClient.ExecuteStreamingSqlStream()
+                    // since it's an abstract class.
```

Comment:
I couldn't find a type to return an empty result here so had to resort to null.

Reply:
You could find the concrete class involved. Personally I generally avoid using mocks, preferring fakes - but I can see how a mock is possibly simpler here.

The lambda expression can be just `(request, _) => null` though.